arch/arm/core/swap.c

```diff
@@ -54,6 +54,9 @@ int arch_swap(unsigned int key)
 	irq_unlock(key);
 #endif
 
+#ifdef CONFIG_NEWLIB_LIBC
+	_impure_ptr = &_current->base.k_reent;
+#endif
```

Comment:
Note that `arch_swap` is only used for co-operative task switching. When a task is preempted, this function is not called.

`_impure_ptr` should be updated in `z_arm_pendsv`, which is the actual common task switching function.

Line 228 of `swap_helper.S` would be a good place to add this.

https://github.com/zephyrproject-rtos/zephyr/blob/816d2c47a4d5dad73fa1fab16b3602e1562d4cda/arch/arm/core/swap_helper.S#L225-L229

Reply:
do we need to make this change in the context switch code for all arches or is there something special about ARM that _impure_ptr needs to be updated like this?

Reply:
> do we need to make this change in the context switch code for all arches

Yes, `struct reent` is basically per-thread context for newlib.

Reply:
> also what about SMP?

For ARM, SMP is not supported at the moment. I will look into this in the future.

Reply:
these changes impact all arches so I think we will need to close on this before it can be merged.

I don't see how a global _impure_ptr updated on context switch could ever work in an SMP system, surely newlib has something for this...does this need to be stored in threa-local storage?

Reply:
perhaps we need to override __getreent()?

Reply:
> perhaps we need to override __getreent()?

Looks like that would be the correct approach for SMP.

https://github.com/eblot/newlib/blob/2a63fa0fd26ffb6603f69d9e369e944fe449c246/newlib/libc/sys/linux/linuxthreads/getreent.c#L5-L10

One problem I see is that `__getreent` is not declared `__weak`, so it would not be override-able.

https://github.com/eblot/newlib/blob/2a63fa0fd26ffb6603f69d9e369e944fe449c246/newlib/libc/reent/getreent.c#L10-L14

I wonder Zephyr should create a separate fork of newlib for this.